File: ride-service/src/main/java/com/example/ride_service/config/KafkaConfig.java
package com.example.ride_service.config;

import com.example.ride_service.dto.RatingIdEvent;
import com.example.ride_service.dto.RideCreatedEvent;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.core.ProducerFactory;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.support.serializer.JsonDeserializer;
import org.springframework.kafka.support.serializer.JsonSerializer;

import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaConfig {

    @Bean
    public ProducerFactory<String, RideCreatedEvent> rideEventProducerFactory() {
        Map<String, Object> config = new HashMap<>();
        config.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092");
        config.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        config.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);
        config.put(JsonSerializer.ADD_TYPE_INFO_HEADERS, false);
        return new DefaultKafkaProducerFactory<>(config);
    }

    @Bean
    public KafkaTemplate<String, RideCreatedEvent> rideEventKafkaTemplate() {
        return new KafkaTemplate<>(rideEventProducerFactory());
    }

    @Bean
    public ConsumerFactory<String, RatingIdEvent> ratingIdConsumerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092");
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "ride-rating-group");
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        props.put(JsonDeserializer.TRUSTED_PACKAGES, "com.example.ride_service.dto");
        props.put(JsonDeserializer.VALUE_DEFAULT_TYPE, RatingIdEvent.class.getName());

        return new DefaultKafkaConsumerFactory<>(
                props,
                new StringDeserializer(),
                new JsonDeserializer<>(RatingIdEvent.class, false)
        );
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, RatingIdEvent>
    ratingIdListenerContainerFactory() {

        ConcurrentKafkaListenerContainerFactory<String, RatingIdEvent> factory =
                new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(ratingIdConsumerFactory());
        return factory;
    }
}

File: rating-service/src/main/kotlin/com/example/rating_service/config/KafkaConfig.kt
package com.example.rating_service.config

import com.example.rating_service.dto.RatingIdInRideEvent
import com.example.rating_service.dto.UserRatingEvent
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import org.springframework.kafka.support.serializer.JsonSerializer

@Configuration
class KafkaConfig {

    // Общая конфигурация для всех продюсеров
    private fun producerFactoryConfig(): Map<String, Any> = mapOf(
        ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to "kafka:9092",
        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to StringSerializer::class.java,
        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to JsonSerializer::class.java,
        JsonSerializer.ADD_TYPE_INFO_HEADERS to false
    )

    @Bean
    fun ratingIdInRideProducerFactory(): ProducerFactory<String, RatingIdInRideEvent> {
        return DefaultKafkaProducerFactory(producerFactoryConfig())
    }

    @Bean
    fun userRatingProducerFactory(): ProducerFactory<String, UserRatingEvent> {
        return DefaultKafkaProducerFactory(producerFactoryConfig())
    }

    @Bean
    fun ratingIdInRideKafkaTemplate(): KafkaTemplate<String, RatingIdInRideEvent> {
        return KafkaTemplate(ratingIdInRideProducerFactory())
    }

    @Bean
    fun userRatingKafkaTemplate(): KafkaTemplate<String, UserRatingEvent> {
        return KafkaTemplate(userRatingProducerFactory())
    }
}